### PR TITLE
JENKINS-51440 Configure alternative context for each Tomcat container

### PR DESCRIPTION
**Pull request by motofix**
Following https://issues.jenkins-ci.org/browse/JENKINS-51440 enhancement request.

### CODE REVIEW

File: src/main/java/hudson/plugins/deploy/DeploymentContext.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package hudson.plugins.deploy;
```
Comment (jansohn) [inline src/main/java/hudson/plugins/deploy/DeploymentContext.java]: This seems overkill. Please keep changes to a minimum (a final String is IMHO enough in this case).

File: src/main/java/hudson/plugins/deploy/NonBlankString.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package hudson.plugins.deploy;
```
Comment (jansohn) [inline src/main/java/hudson/plugins/deploy/NonBlankString.java]: This seems overkill. Please keep changes to a minimum (a final String is IMHO enough in this case).

File: src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java [hunk 3]
```diff
@@ -28,18 +33,24 @@
...
 
-    public TomcatAdapter(String url, String credentialsId) {
-        super(credentialsId);
-        this.url = url;
-        this.path = null;
-    }
+    /**
+     * Alternative context that override context defined in plugin main configuration
+     */
+    public final DeploymentContext alternativeDeploymentContext;
```
Comment (jansohn) [inline src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java]: This seems overkill. Please keep changes to a minimum (a final String is IMHO enough in this case).

File: src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java [hunk 4]
```diff
@@ -28,18 +33,24 @@
...
 
-    public TomcatAdapter(String url, String credentialsId, String path) {
+    private final String path;
+
+    public TomcatAdapter(
+        String url, String credentialsId, String alternativeDeploymentContext, String path) {
         super(credentialsId);
         this.url = url;
         this.path = path;
+        if(StringUtils.isNotBlank(alternativeDeploymentContext)) {
```
Comment (jansohn) [inline src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java]: Please use ternary if.

File: src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java [hunk 5]
```diff
@@ -81,4 +92,15 @@ public void configure(Configuration config, EnvVars envVars, VariableResolver<St
     protected WAR createWAR(File deployableFile) {
         return new TomcatWAR(deployableFile.getAbsolutePath());
     }
+
+    @Override
+    public void redeployFile(
+        FilePath war, String aContextPath, Run<?, ?> run, Launcher launcher, TaskListener listener)
+        throws IOException, InterruptedException {
+        String finalContextPath = aContextPath;
+        if(this.alternativeDeploymentContext != null) {
```
Comment (jansohn) [inline src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java]: Please use ternary if.

File: src/test/java/hudson/plugins/deploy/RemoteCallableTest.java [hunk 6]
```diff
@@ -73,7 +71,7 @@ public void testCallableSerialization () throws Exception {
                 new UsernamePasswordCredentialsImpl(CredentialsScope.GLOBAL, "test-id", "", "user", "pass"));
 
         ArrayList<ContainerAdapter> adapters = new ArrayList<ContainerAdapter>();
-        adapters.add(new Tomcat8xAdapter(j.getURL().toExternalForm(), "test-id", "/manager/text"));
+        adapters.add(new Tomcat8xAdapter("http://example.com", "test-id", null, "/manager/text"));
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/RemoteCallableTest.java]: Don't change the deployment URL back to an existing domain!

File: src/test/java/hudson/plugins/deploy/tomcat/Tomcat6xAdapterTest.java [hunk 7]
```diff
@@ -1,38 +1,36 @@
...
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.io.ByteArrayOutputStream;
+import java.util.Collections;
 
 /**
  * @author frekele
  */
-public class Tomcat6xAdapterTest {
+public class Tomcat6xAdapterTest extends TomcatAdapterTest {
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/tomcat/Tomcat6xAdapterTest.java]: Please keep the changes at a minimum. Refactoring can be done in a separate PR!

File: src/test/java/hudson/plugins/deploy/tomcat/Tomcat7xAdapterTest.java [hunk 8]
```diff
@@ -31,7 +33,7 @@
 /**
  * @author soudmaijer
  */
-public class Tomcat7xAdapterTest {
+public class Tomcat7xAdapterTest extends TomcatAdapterTest {
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/tomcat/Tomcat7xAdapterTest.java]: Please keep the changes at a minimum. Refactoring can be done in a separate PR!

File: src/test/java/hudson/plugins/deploy/tomcat/Tomcat8xAdapterTest.java [hunk 9]
```diff
@@ -34,10 +29,14 @@
...
 
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.util.Collections;
+
 /**
  * @author frekele
  */
-public class Tomcat8xAdapterTest {
+public class Tomcat8xAdapterTest extends TomcatAdapterTest {
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/tomcat/Tomcat8xAdapterTest.java]: Please keep the changes at a minimum. Refactoring can be done in a separate PR!

File: src/test/java/hudson/plugins/deploy/tomcat/TomcatAdapterTest.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package hudson.plugins.deploy.tomcat;
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/tomcat/TomcatAdapterTest.java]: Please keep the changes at a minimum. Refactoring can be done in a separate PR!

File: src/main/resources/hudson/plugins/deploy/tomcat/TomcatAdapter/config.jelly [hunk 11]
```diff
@@ -8,6 +8,9 @@
     <f:textbox />
   </f:entry>
   <f:advanced>
+    <f:entry title="Alternative deployment context" field="alternativeDeploymentContext">
```
Comment (jansohn) [inline src/main/resources/hudson/plugins/deploy/tomcat/TomcatAdapter/config.jelly]: Help file is missing (see https://github.com/jenkinsci/deploy-plugin/blob/master/src/main/resources/hudson/plugins/deploy/tomcat/TomcatAdapter/help-path.html for an example).

File: src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java (hunk 4 above)
Comment (motofix) [reply src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java]: I would like to use `StringUtils.defaultIfBlank`

File: src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java (hunk 5 above)
Comment (motofix) [reply src/main/java/hudson/plugins/deploy/tomcat/TomcatAdapter.java]: I would like to use `StringUtils.defaultIfBlank`

File: src/test/java/hudson/plugins/deploy/tomcat/Tomcat7xAdapterTest.java [hunk 12]
```diff
@@ -86,7 +86,9 @@ public void testVariables() throws Exception {
                 "", getVariable(usernameVariable), password);
         CredentialsProvider.lookupStores(jenkinsRule.jenkins).iterator().next().addCredentials(Domain.global(), c);
 
-        adapter = new Tomcat7xAdapter(getVariable(urlVariable), c.getId(), managerContextPath);
+        adapter =
+            new Tomcat7xAdapter(
+                getVariable(urlVariable), c.getId(), getVariable(alternativeContextVariable), managerContextPath);
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/tomcat/Tomcat7xAdapterTest.java]: You need to add an environment variable if you want to use the `getVariable()` method, e.g. `envVars.put(alternativeContextVariable, alternativeContext);`.

File: src/test/java/hudson/plugins/deploy/tomcat/Tomcat8xAdapterTest.java [hunk 13]
```diff
@@ -91,7 +91,9 @@ public void testVariables() throws Exception {
                 "", getVariable(usernameVariable), password);
         CredentialsProvider.lookupStores(jenkinsRule.jenkins).iterator().next().addCredentials(Domain.global(), c);
 
-        adapter = new Tomcat8xAdapter(getVariable(urlVariable), c.getId(), managerContextPath);
+        adapter =
+            new Tomcat8xAdapter(
+                getVariable(urlVariable), c.getId(), getVariable(alternativeContextVariable), managerContextPath);
```
Comment (jansohn) [inline src/test/java/hudson/plugins/deploy/tomcat/Tomcat8xAdapterTest.java]: You need to add an environment variable if you want to use the `getVariable()` method, e.g. `envVars.put(alternativeContextVariable, alternativeContext);`.

**Pull request merged**